packages/bsky/src/api/blob-resolver.ts

```diff
@@ -1,160 +1,307 @@
-            'blob resolution timeout',
-          )
-          return next(createError(504)) // Gateway timeout
-        }
-        if (!err.response || err.response.status >= 500) {
-          log.warn(
-            { host: err.request?.host, path: err.request?.path },
-            'blob resolution failed upstream',
-          )
-          return next(createError(502))
-        }
-        return next(createError(404, 'Blob not found'))
-      }
-      if (err instanceof DidNotFoundError) {
-        return next(createError(404, 'Blob not found'))
+      if (res.headersSent) {
+        res.destroy()
+      } else {
+        next(err || createError(500))
       }
-      return next(err)
     }
-  })
+  }
+}
 
-  return router
+export function parseBlobParams(params: { cid: string; did: string }) {
+  const { cid, did } = params
+  if (!isAtprotoDid(did)) throw createError(400, 'Invalid did')
+  const cidObj = parseCid(cid)
+  if (!cidObj) throw createError(400, 'Invalid cid')
+  return { cid: cidObj, did }
 }
 
-export async function resolveBlob(ctx: AppContext, did: string, cid: CID) {
+export async function getBlobUrl(
+  dataplane: DataPlaneClient,
+  did: string,
+  cid: CID,
+): Promise<URL> {
+  const pds = await getBlobPds(dataplane, did, cid)
+
+  const url = new URL(`/xrpc/com.atproto.sync.getBlob`, pds)
+  url.searchParams.set('did', did)
+  url.searchParams.set('cid', cid.toString())
+
+  return url
+}
+
+export async function getBlobPds(
+  dataplane: DataPlaneClient,
+  did: string,
+  cid: CID,
+): Promise<string> {
   const cidStr = cid.toString()
 
-  const [identity, { takenDown }] = await Promise.all([
-    ctx.dataplane.getIdentityByDid({ did }).catch((err) => {
-      if (isDataplaneError(err, Code.NotFound)) {
-        return undefined
-      }
-      throw err
-    }),
-    ctx.dataplane.getBlobTakedown({ did, cid: cid.toString() }),
-  ])
-  const services = identity && unpackIdentityServices(identity.services)
-  const pds =
-    services &&
-    getServiceEndpoint(services, {
-      id: 'atproto_pds',
-      type: 'AtprotoPersonalDataServer',
-    })
-  if (!pds) {
-    throw createError(404, 'Origin not found')
-  }
+  const abortController = new AbortController()
+  const { signal } = abortController
+
+  // Start both promises in parallel, but don't use Promise.all as we don't care
+  // about the identity result if the blob was takendown, and we want to return
+  // (or throw) ASAP.
+  const takendownPromise = dataplane.getBlobTakedown({ did, cid: cidStr })
+  const identityPromise = dataplane.getIdentityByDid({ did }, { signal })
+
+  // Avoid "uncaught promise rejection" error (we will await later, if we care)
+  identityPromise.catch(() => {})
+
+  const { takenDown } = await takendownPromise
+  // If the blob is taken down, don't wait for the identity promise
   if (takenDown) {
+    abortController.abort() // We don't care about the identity response
```

Comment:
The amount of time we hit this takedown path, let alone on an identity that isn't known/cached, should be quite tiny!  At a glance this seems a bit unrelated from the axios move—was there something specific on your mind with this change?

Reply:
Reverted to the previous behavior